enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1120 @@
+	// otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.
+	// The value of this field must be no more than 16384 characters in length.
+	// This limit was selected as most common web servers have a limit of 16384 characters
+	// or some lower limit.
+	//
+	// Note that the total size of all added headers *after* interpolating dynamic values
+	// must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
+	// IngressController.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=16384
+	Value string `json:"value" protobuf:"bytes,1,opt,name=value"`
+}
+```
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.
+The configuration template defines a frontend named "public" for non-TLS HTTP connections, as well as frontends named "fe_no_sni" and "fe_sni" for TLS connections for which the router has terminated TLS (viz. for edge-terminated or reencrypt routes).  The template uses
+environment variables as input parameters.
+
+The enhancement adds two new
+environment variables, `ROUTER_HTTP_RESPONSE_HEADERS` and `ROUTER_HTTP_REQUEST_HEADERS`, which specify the request and response HTTP headers to be set or deleted in the
+`public`, `fe_no_sni`, and `fe_sni` frontends.
+
+The ingress operator sets this variable based on the IngressController's
+`spec.httpHeaders.actions.response[*].set` or `spec.httpHeaders.actions.response[*].delete` or `spec.httpHeaders.actions.request[*].set` or `spec.httpHeaders.actions.request[*].delete` field values. 
+
+The value of `ROUTER_HTTP_RESPONSE_HEADERS` or `ROUTER_HTTP_REQUEST_HEADERS` is a comma-delimited list of percent encoded  triplets of the form `name:value:action` for action `Set` and `name:action` pairs for action `Delete`.
+The template translates these values into the appropriate `http-response set-header` stanzas.
+
+For example, if `ROUTER_HTTP_RESPONSE_HEADERS` has percent encoded value `"X-Frame-Options%3ADENY%3ASet%2CX-XSS-Protection%3A1%253Bmode%253Dblock%3ASet%2Cx-forwarded-client-cert%3A%2525%257B%252BQ%257D%255Bssl_c_der%252Cbase64%255D%3ASet%2CX-Frame-Options%3ADelete%2CX-XSS-Protection%3ADelete"`
+, this gets decoded into the 5 triples ("X-Frame-Options", "DENY", "Set"), ("X-XSS-Protection", "1;mode=block", "Set"), ("x-forwarded-client-cert", "%{+Q}[ssl_c_der,base64]", "Set"), ("X-Frame-Options", "Delete"), and ("X-XSS-Protection", "Delete"); and if `ROUTER_HTTP_REQUEST_HEADERS` has the percent encoded value `"Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete"`,
```

Comment:
Wait a minute—are you _doubly_ percent-encoding these strings?!
```sh
% python3 -c 'import urllib.parse, sys; print(urllib.parse.unquote_plus(sys.argv[1]))' 'X-Frame-Options%3ADENY%3ASet%2CX-XSS-Protection%3A1%253Bmode%253Dblock%3ASet%2Cx-forwarded-client-cert%3A%2525%257B%252BQ%257D%255Bssl_c_der%252Cbase64%255D%3ASet%2CX-Frame-Options%3ADelete%2CX-XSS-Protection%3ADelete'
X-Frame-Options:DENY:Set,X-XSS-Protection:1%3Bmode%3Dblock:Set,x-forwarded-client-cert:%25%7B%2BQ%7D%5Bssl_c_der%2Cbase64%5D:Set,X-Frame-Options:Delete,X-XSS-Protection:Delete
% python3 -c 'import urllib.parse, sys; print(urllib.parse.unquote_plus(sys.argv[1]))' 'Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete'
Accept:text%2Fplain%2C+text%2Fhtml:Set,Accept-Encoding:Delete
```
Why would you _double_-encode each string?  I expected the header name and header value to be percent-encoded, but encoding the entire tuple _after_ the header name and value are already encoded seems unnecessarily complicated because then your algorithm must unquote, split, and then unquote again:
```python3
unquoted_input = urllib.parse.unquote(input)
quoted_tuples = unquoted_input.split(",")
for quoted_tuple in quoted_tuples:
    parts = quoted_tuple.split(":")
    if parts[len(parts)-1] == "Set":
        tuple = ((urllib.parse.unquote_plus(parts[0]), urllib.parse.unquote_plus(parts[1]), parts[2]))
    elif parts[len(parts)-1] == "Delete":
        tuple = ((urllib.parse.unquote_plus(parts[0]), parts[1]))
```
If you singly percent-encoded values, then the algorithm would be simpler:
```python3
tuples = input.split(",")         
for tuple in tuples:              
    parts = tuple.split(":")       
    if parts[len(parts)-1] == "Set":
        tuple = ((urllib.parse.unquote(parts[0]), urllib.parse.unquote(parts[1]), parts[2]))          
    elif parts[len(parts)-1] == "Delete":
        tuple = ((urllib.parse.unquote(parts[0]), parts[1]))     
```
And the encoded values would be easier to read—compare:
* `Accept:text%2Fplain%2C%20text%2Fhtml:Set,Accept-Encoding:Delete`
* `Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete`

Reply:
i made the code changes.

Reply:
done